### ISSUE TITLE
Refactoring put_data

### ISSUE THREAD

**Issue by jhamman** [UW-Hydro/VIC#281]
`put_data.c` is in need of a refactor.  The refactor is needed for the following reasons:

1. Currently, `out_data`, `save_data`, and a few other counters are initialized by calling `put_data()` with the `rec = -nrecs`. We end up going through a lot of computations that are never used and we expose a bunch of opportunities to use uninitialized values.  Today, I was running into a case with a divide by zero. Our first step should be separating out this initialization step.
2. Supporting multiple time averaging streams (e.g. output file 1 has daily averages, output file 2 has monthly averages, and output file 3 has 6 hour instantaneous values), requires separating the aggregation step (where we calculate/collect output variables) from the time averaging step.  This is not currently possible.
3. For the `image` and `cesm` drivers, the ways the counters in `calc_water_balance_error` and `calc_energy_balance_error` are currently use are meaningless because they end up being the sum of all grid cells in the `local_domain`. These counters should become output variables. 

My proposal is to greatly simplify `put_data`. Factoring out a number of the tasks that it currently performs.  The proposed data flow would look something like this:

1. `put_data()` - Create output variables.  E.g. `out_data[OUT_EVAP_BARE] = cell.layer[index].evap *    cell.layer[index].bare_evap_frac`.
2. `agg_subgrid_data()` - generic functions for aggregating subgrid tiles.  These should remain separate from `put_data()` so that we can, in the future, create output files with the full subgrid matrix.
3. `agg_temporal_data()` - time aggregation / averaging.  
4. `save_data()` - write to disk

The first three will live in `driver/shared` while the fourth will be driver specific.  

xref: #196, #198

**Comment by jhamman** [comment]
I should add, the user interface to the end goal (beyond this first issue) here may look something like this:

```BASH
# DAILY AVERAGES
OUTFILE 1
HIST_FREQ -24
RECS_PER_FILE 1
DIURNAL_BINS_PER_DAY 1
INCLUDE_SUBGRID NONE
OUT_RUNOFF
OUT_SURF_TEMP
...
OUT_EVAP

# MONTHLY MEAN 3HRLY DIURNAL CYCLE
OUTFILE 2
HIST_FREQ 0
RECS_PER_FILE 8
DIURNAL_BINS_PER_DAY 8
INCLUDE_SUBGRID SNOW_BANDS
OUT_LATENT
OUT_SENSIBLE
...
OUT_GRND_FLUX
```

A summary of the configuration parameters is below: 

| Parameter       | Type (values)                                     | Description                                                           | Note             |
|-----------------|---------------------------------------------------|-----------------------------------------------------------------------|------------------|
| OUTFILE         | integer [1, MAX_OUT_FILES)                        | history stream identifier                                             | Required         |
| HIST_FREQ       | integer {>0 = timesteps, <0 = hours, 0 = monthly} | history file timestep                                                 | default =-24     |
| RECS_PER_FILE   | integer [1, MAX_OUT_FILES)                        | number of history timesteps to put in each file                       | default = 1      |
| DIURNAL_STEPS   | integer [1, 72]                                   | number of diurnal timesteps to include (only used when HIST_FREQ = 0) | default = 1      |
| INCLUDE_SUBGRID | string {all, snow_bands, veg_tiles, none}         | types of subgrid aggregation available to write                       | default = none   |
| AGG_TYPE        | string {avg, min, max, inst, var}                 | type of temporal aggregation                                          | default = avg    |
| PRECISION       | string {single, double}                           | output file precision                                                 | default = double |

**Comment by bartnijssen** [comment]
I am very much in favor of this. I'll have to read the proposal itself in somewhat more detail. I would also suggested that we call out the initialization step separately rather than passing a negative number of nrecs.

**Comment by jhamman** [comment]
closed via #494
